File: flatjson.js
#!/usr/bin/env node
const fs = require('fs');

let quoted = (s) => s;
let quotedKey = quoted;

function write(s)
{
  const len = fs.writeSync(1, s);
  if (len != s.length)
    throw 42;
}

const printJSON = function(data, prefix)
{
  switch (data.constructor.name)
  {
    case 'Array':
      for (const i in data)
        printJSON(data[i], `${prefix}[${i}]`);
      break;
    case 'Object':
      for (const k in data)
        printJSON(data[k], `${prefix}.${quotedKey(k)}`);
      break;
    case 'String':
      write(`${prefix} = ${quoted(data)}\n`);
      break;
    default:
      write(`${prefix} = ${data}\n`);
  }
}

const initQuoted = function(f)
{
  quoted = f;
  const pattern = /[\x00-\x2F\x3A-\x40\x5B-\x60\x7E\x7F]/;
  quotedKey = (s) => pattern.test(s) ? f(s) : s;
}

let ifile = 2;

switch (process.argv[2]) {
  case '-h':
  case '--help':
    console.error(`Usage: flatjson [-q | -Q] [FILE]...
 -q          string are in single quote
 -Q          string are in double quote
 -h, --help  this help`);
    break;

  case '-q': {
    const replacements = {
      '\'': '\\\'',
      '\\': '\\\\',
      '\a': '\\a',
      '\b': '\\b',
      '\t': '\\t',
      '\n': '\\n',
      '\v': '\\v',
      '\f': '\\f',
      '\r': '\\r',
    };
    const reg = /['\\\x07-\x0D]/g
    initQuoted((s) => `'${s.replace(reg, (c) => replacements[c])}'`);
    ++ifile;
    break;
  }

  case '-Q': {
    const replacements = {
      '\"': '\\\"',
      '\\': '\\\\',
      '\a': '\\a',
      '\b': '\\b',
      '\t': '\\t',
      '\n': '\\n',
      '\v': '\\v',
      '\f': '\\f',
      '\r': '\\r',
    };
    const reg = /["\\\x07-\x0D]/g
    initQuoted((s) => `"${s.replace(reg, (c) => replacements[c])}"`);
    ++ifile;
    break;
  }
}

let filename = process.argv[ifile] || '/dev/stdin';
let errcode = 0;

const processFile = function(err, data) {
  if (err) {
    console.error(err.toString());
    errcode = errcode || 2;
  }
  else {
    let json;
    try {
      json = JSON.parse(data);
    }
    catch (e) {
      console.error(`Parse error: ${filename}: ${e.message}`)
      errcode = errcode || 3;
    }

    try {
      printJSON(json, '');
    }
    catch (e) {
      process.exit(errcode || 4);
    }
  }

  ++ifile;
  if (ifile < process.argv.length) {
    filename = process.argv[ifile];
    fs.readFile(filename, 'utf8' , processFile);
  }
  else {
    process.exit(errcode);
  }
}

fs.readFile(filename, 'utf8', processFile);

File: flatjson.py
#!/usr/bin/env python3
from sys import stdout, argv, stderr
from json import load as json_load, JSONDecodeError
import re
import sys

# for big json
sys.setrecursionlimit(2000)

def quoted(s: str) -> str:
  return s

quotedKey = quoted


def print_json(data, prefix: str = '') -> None:
  t = type(data)
  if t == dict:
    for k, v in data.items():
      print_json(v, f'{prefix}.{quotedKey(k)}')
  elif t == list:
    for i, v in enumerate(data):
      print_json(v, f'{prefix}[{i}]')
  elif t == str:
    stdout.write(f'{prefix} = {quoted(data)}\n')
  else:
    stdout.write(f'{prefix} = {data}\n')


def makeQuoted(c: str):
  replacements = {
    c: '\\'+c,
    '\\': '\\\\',
    '\a': '\\a',
    '\b': '\\b',
    '\t': '\\t',
    '\n': '\\n',
    '\v': '\\v',
    '\f': '\\f',
    '\r': '\\r',
  }
  patt = re.compile(fr"[{c}\\\x07-\x0D]");
  pattCheck = re.compile(r"^\w*$")
  return (lambda s: f'{c}{patt.sub(lambda m: replacements[m[0]], s)}{c}',
          lambda s: s if pattCheck.match(s) else quoted(s))


ifile = 1

if len(argv) > 1:
  arg = argv[1]

  if arg == '-h' or arg == '--help':
    print('''Usage: flatjson [-q | -Q] [FILE]...
 -q          string are in single quote
 -Q          string are in double quote
 -h, --help  this help''')
    exit(0)

  elif arg == '-q':
    quoted, quotedKey = makeQuoted("'")
    ifile += 1;

  elif arg == '-Q':
    quoted, quotedKey = makeQuoted('"')
    ifile += 1

  elif arg == '--':
    ifile += 1


errcode = 0

if len(argv) <= ifile:
  ifile -= 1
  argv[ifile] = '/dev/stdin'

for i in range(ifile, len(argv)):
  try:
    with open(argv[i], 'rb') as f:
      try:
        json = json_load(f)
      except JSONDecodeError:
        errcode = errcode or 2
        continue

      try:
        print_json(json)
      except OSError as e:
        exit(errcode)

  except OSError:
    errcode = errcode or 1
exit(errcode)
